### ISSUE TITLE
Numpy functions return a scalar when the input is a 0-dimensional array

### ISSUE THREAD

**Issue by evertrol** [numpy/numpy#6051]
(This may be a duplicate issue, but so far I've only found few somewhat related issues, e.g. #3161 , #2500 and perhaps #1774 and #3311.)

When creating an array of a single (non-list) element, the result is a 0-dimensional array:

    >>> a = np.array(1)
    >>> type(a), a.shape, a.ndim
    (<class 'numpy.ndarray'>, (), 0)

When applying a numpy function to `a`, the result is a numpy scalar instead:

    >>> b = np.cos(a)
    >>> type(b), b.shape, b.ndim
    (<class 'numpy.float64'>, (), 0)

(or one can simply multiply by an int:

    >>> b = 100 * a
    >>> type(b), b.shape, b.ndim
    (<class 'numpy.int64'>, (), 0)
)
This effectively changes the type, and options available to the result. For example, one can assign to an indexed 0-dimensional array, but not to an indexed numpy scalar:

    >>> a[True] = 5
    >>> b[True] = 5
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    TypeError: 'numpy.float64' object does not support item assignment

Of course, indexing a scalar is awkward, but it is consistent if one doesn't known whether the input is a (Python) float or int (to which (as)array has been applied), a list, a one element array `[1]` or a 1 or higher-dimensional array: code written would not need to distinguish between a scalar and an array.

My workaround for now is to use `asarray`:

    >>> b = np.asarray(b)
    >>> b[True] = 5

but I wonder if there is a better way.

Moreover, how consistent should the return type of numpy functions be with the input type?

Apologies if this topic has come up before, and may have been discussed to dead, but there was relatively little I could find that appears to address exactly this issue.

**Comment by seberg** [comment]
This is true, and personally I agree (it probably is a duplicate, but I don't know of what). The thing is that especially all ufuncs (so all elementwise functions, which are the  backbone of numpy really) will convert 0-d arrays to scalar for the result and not remember the input as you mentioned. Since we convert everything to arrays on input (also scalars), you do not remember that it was a scalar. You could say that currently you cannot distinguish/remember `shape=None` (scalar) from `shape=()` (array), so you go with scalar, since it is the safer bet.
Right now, I would personally be for changing this for ufuncs. For reductions, it is also pretty simple since `arr.sum(axis=None)` is scalar, but `arr.sum(axis=range(arr.ndim))` would be an array....

However, I am neither sure it can be done easily, nor sure about compatibility issues. Also we might want to refactor scalars completely (they duplicate loads of code currently) in the longer run, though I do not think that should stop anyone from thinking about attacking this issue.

**Comment by eric-wieser** [comment]
Somewhat related to gh-13105

**Comment by mdhaber** [comment]
If scalars are around for good, and if NumPy functions will continue to return scalars instead of 0-D arrays, can scalars support item issignment?

**Comment by seberg** [comment]
> can scalars support item issignment?

I don't see how that would help.  What we can do is return 0-D arrays when the input is 0-D, I have been chipping away at making that step at least possible.  That step removes the need for the `asarray()` unless you already don't know what the input is.
The next thing we can do for some things (like ufuncs) is to allow an `out=` sentinel to say "must return array".  I got stuck on that a long while back, but the same "chipping away" may make that much saner now as well.

**Comment by seberg** [comment]
Forgot to link: gh-24897 which has more relevant discussion (and which is really a duplicate).

**Comment by mdhaber** [comment]
Ah, great. That is better; I thought that idea was dead based on the text in NEP 56.

**Comment by seberg** [comment]
> I thought that idea was dead based on the text in NEP 56.

No, it just came up once or twice and I wanted to point out that while maybe related it is a different issue and if you try to fix everything at once you end up fixing nothing.
